### ISSUE TITLE
Mac OS Cannot use syntax checker python-flake8 in this buffer

### ISSUE THREAD

**Issue by zhangjie2012** [flycheck/flycheck#1437]
I had install `flake8` and can run flake8 in eshell. 

`flycheck-verify-setup` show :  

```
Syntax checkers for buffer test.py in python-mode:

  python-flake8 (disabled)
    - may enable:         Automatically disabled!
    - executable:         Found at /usr/bin/python
    - configuration file: Not found
    - `flake8' module:    Missing

  python-pylint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Found at /usr/bin/python
    - configuration file: Found at "/Users/luanhua/.pylintrc"
    - `pylint' module:    Missing

  python-pycompile
    - may enable: yes
    - executable: Found at /usr/bin/python

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.
```

`C-c ! C-c` choose `python-flake8` show `Cannot use syntax checker python-flake8 in this buffer`.

what's wrong ?

My config file work well on ubuntu.

**Comment by cpitclaudel** [comment]
You probably installed flake8 for python 3, but your buffer is in python 2?
Try either changing flycheck-python-flake8-executable (to point to python3), or installing flake8 with pip2

**Comment by noxdafox** [comment]
I have a similar issue when trying to use `python-pylint`.

```
Syntax checker in buffer factory.py in python-mode:

  python-pylint (disabled)
    - major mode:         `python-mode' supported
    - may enable:         Automatically disabled!
    - executable:         Found at /usr/bin/pylint3
    - configuration file: Found at "/home/noxdafox/.pylintrc"
    - `pylint' module:    Missing

Flycheck cannot use this syntax checker for this buffer.

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20180326.836)
Emacs version:    25.2.2
System:           x86_64-pc-linux-gnu
Window system:    x
```

I can't enable `python-pylint` on any Python buffer. Yet Pylint3 is correctly installed and functional.

**Comment by zhangjie2012** [comment]
@noxdafox  hey,  `(setq flycheck-python-flake8-executable "flake8")` fix it .

**Comment by zhangjie2012** [comment]
@noxdafox hey, I use (setq flycheck-python-flake8-executable "flake8") fix it .

**Comment by noxdafox** [comment]
I already have set `(setq flycheck-python-pylint-executable "pylint3")`. As you can see, the executable path is correctly reported when describing `pyhton-pylint`.

**Comment by cpitclaudel** [comment]
@noxdafox Thanks for the report.  Can you check whether using `(setq flycheck-python-pylint-executable "python3")` works?

It would also be helpful if you could try adding this to your .emacs and seeing if it fixes the problem:

```elisp
(require 'flycheck)
(defun flycheck-python-needs-module-p (checker)
  "Determines whether CHECKER needs to be invoked through Python.
Previous versions of Flycheck called pylint and flake8 directly;
this check ensures that we don't break existing code."
  (not (string-match-p (rx (or "pylint" "flake8" "pylint3")
                           (or "-script.pyw" ".exe" ".bat" "")
                           eos)
                       (flycheck-checker-executable checker))))
```

The issue is that I didn't realize when I made the patch that `pylint3` was a valid name for pylint.

**Comment by noxdafox** [comment]
Setting `(setq flycheck-python-pylint-executable "python3")` seems to fix the issue.

`pylint3` is the executable name used in Debian and derivates. I now realized I did not suffer from this issues because on other machines I set the alias `pylint3 -> pylint` as I mostly develop on Python 3.

**Comment by cpitclaudel** [comment]
Thanks; can you try the other change, too?

**Comment by togakangaroo** [comment]
I am seeing a similar situation. I am using pipenv so I for the sake of figuring things out just wired flake8 directly into `pipenv run which flake8` and set my flakerc directly (it was not being found before)

Syntax checkers for buffer sales_handoff.py in python-mode:

```
  python-flake8 (disabled)
    - may enable:         yes
    - executable:         Found at /Users/gmauer/.local/share/virtualenvs/project-qFaGhtMo/bin/flake8
    - configuration file: Found at "/Users/gmauer/.flake8"

  python-pylint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Found at /Users/gmauer/.pyenv/shims/python3
    - configuration file: Not found
    - `pylint' module:    Missing

  python-pycompile
    - may enable: yes
    - executable: Found at /Users/gmauer/.pyenv/shims/python

Flycheck Mode is enabled. Use SPC u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20180415.1527)
Emacs version:    25.3.1
System:           x86_64-apple-darwin17.3.0
Window system:    ns

```

However, no amount of enabling will turn on the checker and trying to run `flycheck-select-checker` for `python-flake8` Just gives me

    can't use syntax checker python-flake8 in this buffer

The buffer is the standard python-mode configured via spacemacs' python layer with no customization.

**Comment by cpitclaudel** [comment]
@togakangaroo Don't you need `(setq flycheck-python-flake8-executable "python3")` instead?
I think your problem might be an issue with the order in which you configure flycheck variables and start flycheck.  Can you show your config?

**Comment by bitsandscraps** [comment]
Having the same issue. This is the result of running `flycheck-select-checker` and then `python-pylint`.

```
  python-pylint (disabled)
    - major mode:         `python-mode' supported
    - may enable:         yes
    - executable:         Found at /home/wisrl/.local/share/virtualenvs/player_mine_py-Fftm8ZS7/bin/python3
    - configuration file: Found at "/home/wisrl/.pylintrc"
    - `pylint' module:    Found at "/home/wisrl/.local/share/virtualenvs/player_mine_py-Fftm8ZS7/lib/python3.5/site-packages/pylint/__init__.py"

Flycheck cannot use this syntax checker for this buffer.

Flycheck Mode is enabled.  Use M-x universal-argument C-c ! x to
enable disabled checkers.

--------------------

Flycheck version: 32snapshot
Emacs version:    26.1
System:           x86_64-pc-linux-gnu
Window system:    x
```

Here's my config.

```
(use-package flycheck
  :straight t
  :init
  (add-hook 'after-init-hook #'global-flycheck-mode)
  (setq flycheck-python-pylint-executable "python3"))

(use-package flycheck-pos-tip
  :straight t)

(with-eval-after-load 'flycheck
  (flycheck-pos-tip-mode))
```

**Comment by cpitclaudel** [comment]
Are there any cases of this that are not fixed by setting `flycheck-python-flake8-executable` or `flycheck-python-pylint-executable` to `python3`? @wbolster , can you clarify what the problem was?

**Comment by wbolster** [comment]
in the case of my dotfiles, see https://github.com/wbolster/dotfiles/commit/3c5c4a63447b1a3d5094280283d951986e859744

> [emacs] force flycheck to invoke flake8 directly
>
> …instead of having it do ‘smart’ things using ‘python -m ...’ which
> does not work well when not in an active venv, and a global flake8 is
> installed via a tool like pipsi.
>
> see also
>
> - flycheck/flycheck#1055
> - flycheck/flycheck#1437

in my case, i use `pipsi` to install flake8 for ‘global use’, i.e. not inside a project-specific venv. the result is that invoking `flake8` works, while `python -m flake8` does not:

```
$ which flake8
/home/wbolster/.local/bin/flake8

$ python -m flake8
/usr/bin/python: No module named flake8
```

in this setup, `(setq flycheck-python-flake8-executable "flake8")` makes it work as it should.

(and if i _also_ have flake8 inside an active virtualenv, then that one is used. this is good.)

**Comment by cpitclaudel** [comment]
With that strategy, don't you get warnings about missing packages when you have packages installed inside the virtualenv but not outside of it?

**Comment by cpitclaudel** [comment]
Or does flake8 not care about imports?

**Comment by wbolster** [comment]
`flake8` does not care about imports, but even if it did, installing _another_ flake8 inside the virtualenv would be the answer. this is how i use `mypy`, which i installed both via `pipx` (globally), and also locally in some project's virtualenvs.

that way, flycheck will use the one from the virtualenv if it's active (you can use `direnv.el` for it, for instance), and will fall back o the ‘global’ one installed via `pipx` from my home dir. this is exactly what i want.

**Comment by cpitclaudel** [comment]
Thanks @wbolster.  I've pondered this and I'm not sure I see a way to support both switching between python2 and python3 easily and falling back to the global pylint when the current virtualenv doesn't have it. 

But your use case is definitely valid, so I've documented it in the Python section of Flycheck's manual to ensure that we don't break it in the future, and I added a note about virtualenvs.  I've also improved the diagnoses produced when a checker can't be found by including (in Flycheck verification buffers) the Python path that Flycheck is seeing.

(All of this is on branch https://github.com/flycheck/flycheck/tree/cpitclaudel_pylint-executable)
